src/overlay/FlowControl.cpp

```diff
@@ -188,48 +256,91 @@ FlowControl::endMessageProcessing(StellarMessage const& msg,
     ZoneScoped;
 
     mFloodDataProcessed += mFlowControlCapacity->releaseLocalCapacity(msg);
+    if (mFlowControlBytesCapacity)
+    {
+        mFloodDataProcessedBytes +=
+            mFlowControlBytesCapacity->releaseLocalCapacity(msg);
+    }
 
     releaseAssert(mFloodDataProcessed <=
                   mApp.getConfig().FLOW_CONTROL_SEND_MORE_BATCH_SIZE);
     bool shouldSendMore = mFloodDataProcessed ==
                           mApp.getConfig().FLOW_CONTROL_SEND_MORE_BATCH_SIZE;
+    if (mFlowControlBytesCapacity)
+    {
+        shouldSendMore =
+            shouldSendMore ||
+            mFloodDataProcessedBytes >=
+                mApp.getConfig().FLOW_CONTROL_SEND_MORE_BATCH_SIZE_BYTES;
+    }
     auto peerPtr = peer.lock();
 
     if (shouldSendMore && peerPtr)
     {
-        sendSendMore(mFloodDataProcessed, peerPtr);
+        if (mFlowControlBytesCapacity)
+        {
+            sendSendMore(mFloodDataProcessed, mFloodDataProcessedBytes,
+                         peerPtr);
+        }
+        else
+        {
+            sendSendMore(mFloodDataProcessed, peerPtr);
+        }
         mFloodDataProcessed = 0;
+        mFloodDataProcessedBytes = 0;
     }
 }
 
 bool
 FlowControl::canRead() const
 {
-    return mFlowControlCapacity->getCapacity().mTotalCapacity > 0;
+    bool canReadBytes =
+        !mFlowControlBytesCapacity || mFlowControlBytesCapacity->canRead();
+    return canReadBytes && mFlowControlCapacity->canRead();
 }
 
 uint32_t
 FlowControl::getNumMessages(StellarMessage const& msg)
 {
-    releaseAssert(msg.type() == SEND_MORE);
-    return msg.sendMoreMessage().numMessages;
+    return msg.type() == SEND_MORE ? msg.sendMoreMessage().numMessages
+                                   : msg.sendMoreExtendedMessage().numMessages;
 }
 
 bool
 FlowControl::isSendMoreValid(StellarMessage const& msg,
                              std::string& errorMsg) const
 {
-    releaseAssert(msg.type() == SEND_MORE);
-    if (getNumMessages(msg) == 0)
+    bool sendMoreExtendedType =
+        mFlowControlBytesCapacity && msg.type() == SEND_MORE_EXTENDED;
```

Comment:
send more and send more extended, how they are different?

Reply:
send_more is for current fc (messages only), while `extended` includes byte counts